File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/accounts/RefreshToken.java
package com.zappymart.domain.accounts;

import java.time.Instant;
import java.util.Objects;

public record RefreshToken(
        String id,
        String sessionId,
        String tokenHash,
        Instant createdAt,
        Instant expiresAt,
        Instant rotatedAt) {

    public RefreshToken {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(sessionId, "sessionId");
        Objects.requireNonNull(tokenHash, "tokenHash");
        Objects.requireNonNull(createdAt, "createdAt");
        Objects.requireNonNull(expiresAt, "expiresAt");
    }

    public static RefreshToken issued(String id, String sessionId, String tokenHash,
                                      Instant moment, Instant expiresAt) {
        return new RefreshToken(id, sessionId, tokenHash, moment, expiresAt, null);
    }

    public boolean wasAlreadyUsed() {
        return rotatedAt != null;
    }

    public boolean hasExpiredAt(Instant moment) {
        return !moment.isBefore(expiresAt);
    }

    public RefreshToken rotatedAt(Instant moment) {
        return new RefreshToken(id, sessionId, tokenHash, createdAt, expiresAt, moment);
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/accounts/RefreshToken.kt
package nl.zappymart.domain.accounts

import java.time.Instant

data class RefreshToken(
    val tokenHash: String,
    val sessionId: String,
    val issuedAt: Instant,
    val expiresAt: Instant,
    val rotatedAt: Instant?,
) {

    fun isUsableAt(moment: Instant): Boolean = rotatedAt == null && moment.isBefore(expiresAt)

    fun wasAlreadyUsed(): Boolean = rotatedAt != null

    fun rotatedAt(moment: Instant) = copy(rotatedAt = moment)
}
